From: Jan Kara <jack@suse.com>
Subject: cachefiles: Avoid deadlocks with fs freezing
References: bsc#935123
Patch-mainline: Never, fixed differently by 5accdf82ba25a and company

Wait for filesystem to thaw before grabbing s_umount semaphore and syncing
the filesystem to avoid sync blocking on frozen filesystem with s_umount held.

Signed-off-by: Jan Kara <jack@suse.com>

---
 fs/cachefiles/interface.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/fs/cachefiles/interface.c
+++ b/fs/cachefiles/interface.c
@@ -368,7 +368,13 @@ static void cachefiles_sync_cache(struct
 	/* make sure all pages pinned by operations on behalf of the netfs are
 	 * written to disc */
 	cachefiles_begin_secure(cache, &saved_cred);
+retry:
+	vfs_check_frozen(cache->mnt->mnt_sb, SB_FREEZE_WRITE);
 	down_read(&cache->mnt->mnt_sb->s_umount);
+	if (cache->mnt->mnt_sb->s_frozen != SB_UNFROZEN) {
+		up_read(&cache->mnt->mnt_sb->s_umount);
+		goto retry;
+	}
 	ret = sync_filesystem(cache->mnt->mnt_sb);
 	up_read(&cache->mnt->mnt_sb->s_umount);
 	cachefiles_end_secure(cache, saved_cred);
